# Lab 1
**Student: Nguyen Mai Dinh, Le (Demi, Le)**

**ID: 300312139**

# Part 1: Data Loading & Exploration with Pandas

## Load the CSV file into a Pandas DataFrame and display the first 5 rows

In [1]:
import pandas as pd
import numpy as np

#Load dataset into DataFrame
student_df = pd.read_csv("student_performance.csv")

#Display top 5 rows
student_df.head(5)

student_id   age  gender  math_score  reading_score  attendance  \
0           1  17.0    Male          85             78        0.95   
1           2  16.0  Female          92             88        0.89   
2           3   NaN   Other         105             65        0.75   
3           4  18.0    Male          -5             72        0.82   
4           5  19.0  Female          78             81        0.91   

  extracurricular   exam_date  
0             Yes  2023-09-10  
1             Yes  2023-09-11  
2              No  2023-09-12  
3             Yes  2023-09-13  
4              No  2023-09-10

## Explore the Dataset

In [2]:
#Check null values, data types
student_df.info()
#Summarize descriptive statistics
student_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   student_id       22 non-null     int64  
 1   age              21 non-null     float64
 2   gender           19 non-null     object 
 3   math_score       22 non-null     int64  
 4   reading_score    22 non-null     int64  
 5   attendance       20 non-null     float64
 6   extracurricular  22 non-null     object 
 7   exam_date        22 non-null     object 
dtypes: float64(2), int64(3), object(3)
memory usage: 1.5+ KB


student_id        age  math_score  reading_score  attendance
count   22.000000  21.000000   22.000000      22.000000   20.000000
mean    10.590909  14.904762   78.863636      80.318182    0.835500
std      5.687573  10.718698   22.380467       8.648624    0.087688
min      1.000000 -17.000000   -5.000000      65.000000    0.680000
25%      6.250000  17.000000   72.000000      73.500000    0.750000
50%     10.500000  18.000000   83.000000      80.500000    0.855000
75%     14.750000  19.000000   91.750000      87.750000    0.910000
max     20.000000  21.000000  105.000000      95.000000    0.950000

## Questions & Answers

**3.1**

Based on the summary table from info():
- The dataset contains a total of 8 fields.
- The columns age, gender, and attendance have missing values.
- The columns gender, extracurricular, and exam_date MIGHT have inconsistent data since their datatype is object.

Based on the summary table from describe():
- The distributions of age and math_score are left-skewed (mean > median).
- The distributions of reading_score and attendance are likely symmetric (mean ≈ median).
- The columns age and math_score may contain outliers since their minimum values are negative, and the maximum math_score is 105, while ages and scores should not be negative, and scores should not exceed 100.
- Age and math_score have high standard deviations, indicating a wide spread and significant variation from the mean.

**3.2**

The columns age, gender, and attendance have missing values.
- Age: 1 missing value
- Gender: 3 missing values
- Attendance: 2 missing values

**3.3**

- The columns age and math_score may contain outliers since their minimum values are negative, and the maximum math_score is 105, while ages and scores should not be negative, and scores should not exceed 100.

3.4.

# Part 2: Data Cleaning with NumPy & Pandas

## Handle Missing Values

In [3]:
# Age column: 1 missing value
# Since there are outliers in age column, mean value might be misleading. 
# Therefore, median is used as a solution to fill in the missing values that is not affected by the outlier
age_median = student_df['age'].median()
student_df['age'] = student_df['age'].fillna(age_median)
student_df['age']

0     17.0
1     16.0
2     18.0
3     18.0
4     19.0
5     20.0
6     17.0
7     21.0
8    -17.0
9    -17.0
10    16.0
11    17.0
12    18.0
13    19.0
14    20.0
15    18.0
16    16.0
17    17.0
18    18.0
19    19.0
20    20.0
21    21.0
Name: age, dtype: float64

In [4]:
# Gender column: 3 missing values  
# The Gender column is not numeric, so an effective approach to filling in missing values is to use a text value indicating that the data is unknown.  

print("Before cleaning: ", student_df['gender'].unique())

# After checking the unique values in the Series, it is observed that besides "Male" and "Female," there is an existing option, "Other."  
# Using "Other" as the fill-in value maintains data consistency without affecting the overall integrity of the dataset.  
student_df['gender'] = student_df['gender'].fillna("Other")
print("After cleaning: ", student_df['gender'].unique())
student_df['gender']

Before cleaning:  ['Male' 'Female' 'Other' nan]
After cleaning:  ['Male' 'Female' 'Other']


0       Male
1     Female
2      Other
3       Male
4     Female
5      Other
6       Male
7     Female
8       Male
9       Male
10    Female
11      Male
12     Other
13    Female
14      Male
15     Other
16    Female
17      Male
18    Female
19      Male
20     Other
21    Female
Name: gender, dtype: object

In [5]:
#Attendance column: 2 missing values
#Attendance is not affected by outliers, so it is good to use median to fill in missing values

attendance_mean = student_df['attendance'].mean()
student_df['attendance'] = student_df['attendance'].fillna(attendance_mean)
student_df['attendance']

0     0.9500
1     0.8900
2     0.7500
3     0.8200
4     0.9100
5     0.9100
6     0.6800
7     0.8900
8     0.8355
9     0.8355
10    0.9400
11    0.7500
12    0.8200
13    0.7500
14    0.8900
15    0.7700
16    0.6800
17    0.9100
18    0.9400
19    0.8200
20    0.8900
21    0.7500
Name: attendance, dtype: float64

## Remove duplicates

In [6]:
# Check the percentage of duplicates
percentageDuplicates = student_df.duplicated().mean() * 100
print("Percentage of duplicates before removing: ", percentageDuplicates, "%")
# Check the row that has duplicated values
duplicatedRow_index = student_df[student_df.duplicated()].index.tolist()
print("Row index with duplicates: ", duplicatedRow_index)

#Remove duplicates
student_df = student_df.drop_duplicates()

# Check if the duplicates are removed
percentageDuplicates = student_df.duplicated().mean() * 100
print("Percentage of duplicates after removing: ", percentageDuplicates, "%")

Percentage of duplicates before removing:  4.545454545454546 %
Row index with duplicates:  [9]
Percentage of duplicates after removing:  0.0 %


## Handle Outliers

In [7]:
student_df = student_df[(student_df['math_score'] >= 0) & (student_df['math_score'] <= 100)]
student_df

student_id   age  gender  math_score  reading_score  attendance  \
0            1  17.0    Male          85             78      0.9500   
1            2  16.0  Female          92             88      0.8900   
4            5  19.0  Female          78             81      0.9100   
5            6  20.0   Other          88             90      0.9100   
6            7  17.0    Male          65             70      0.6800   
7            8  21.0  Female          95             92      0.8900   
8            9 -17.0    Male          72             68      0.8355   
10          10  16.0  Female          82             85      0.9400   
11          11  17.0    Male          89             76      0.7500   
13          13  19.0  Female          93             87      0.7500   
14          14  20.0    Male          77             73      0.8900   
15          14  18.0   Other          70             90      0.7700   
16          15  16.0  Female          68             82      0.6800   
17          16  17.0    Male          84             79      0.9100   
18          17  18.0  Female          91             88      0.9400   
19          18  19.0    Male          74             75      0.8200   
20          19  20.0   Other          99             95      0.8900   
21          20  21.0  Female          60             85      0.7500   

   extracurricular   exam_date  
0              Yes  2023-09-10  
1              Yes  2023-09-11  
4               No  2023-09-10  
5              Yes  2023-09-11  
6               No  2023-09-12  
7              Yes  2023-09-13  
8               No  2023-09-10  
10             Yes  2023-09-11  
11              No  2023-09-12  
13              No  2023-09-10  
14             Yes  2023-09-11  
15             Yes  2023-09-13  
16              No  2023-09-12  
17             Yes  2023-09-13  
18             Yes  2023-09-10  
19              No  2023-09-11  
20             Yes  2023-09-12  
21              No  2023-09-13

**4.1. Display Min/Max after cleaning**

In [8]:
min_value = student_df['math_score'].min()
max_value = student_df['math_score'].max()
print("Min value after cleaning: ", min_value)
print("Max value after cleaning: ", max_value)

Min value after cleaning:  60
Max value after cleaning:  99


**4.2. Alternative Solution** 

There is another way to detect outliers is by creating a list and iterating through each row to check if the value is outside the valid range.
Then, rows with outlier values can be dropped.

In [9]:
#Create empty list to store outliers
outliers_arr = []

#Iterate through each row to check if the value is in valid range
for value in student_df['math_score']:
    valid_range = range(0, 101)
    if value not in valid_range:
        #if condition is met, append value to the list
        outliers_arr.append(value)

# Make a copy to avoid modifying the original version
student_df_copy = student_df.copy()

#Drop rows that contain outlier in the outlier list
for outlier in outliers_arr:
    student_df_copy = student_df_copy.drop(index=student_df_copy[student_df_copy['math_score'] == outlier].index)

#Check results
min_value_copy = student_df_copy['math_score'].min()
max_value_copy = student_df_copy['math_score'].max()
print("Min value after cleaning: ", min_value_copy)
print("Max value after cleaning: ", max_value_copy)

student_df_copy

Min value after cleaning:  60
Max value after cleaning:  99


student_id   age  gender  math_score  reading_score  attendance  \
0            1  17.0    Male          85             78      0.9500   
1            2  16.0  Female          92             88      0.8900   
4            5  19.0  Female          78             81      0.9100   
5            6  20.0   Other          88             90      0.9100   
6            7  17.0    Male          65             70      0.6800   
7            8  21.0  Female          95             92      0.8900   
8            9 -17.0    Male          72             68      0.8355   
10          10  16.0  Female          82             85      0.9400   
11          11  17.0    Male          89             76      0.7500   
13          13  19.0  Female          93             87      0.7500   
14          14  20.0    Male          77             73      0.8900   
15          14  18.0   Other          70             90      0.7700   
16          15  16.0  Female          68             82      0.6800   
17          16  17.0    Male          84             79      0.9100   
18          17  18.0  Female          91             88      0.9400   
19          18  19.0    Male          74             75      0.8200   
20          19  20.0   Other          99             95      0.8900   
21          20  21.0  Female          60             85      0.7500   

   extracurricular   exam_date  
0              Yes  2023-09-10  
1              Yes  2023-09-11  
4               No  2023-09-10  
5              Yes  2023-09-11  
6               No  2023-09-12  
7              Yes  2023-09-13  
8               No  2023-09-10  
10             Yes  2023-09-11  
11              No  2023-09-12  
13              No  2023-09-10  
14             Yes  2023-09-11  
15             Yes  2023-09-13  
16              No  2023-09-12  
17             Yes  2023-09-13  
18             Yes  2023-09-10  
19              No  2023-09-11  
20             Yes  2023-09-12  
21              No  2023-09-13

In [10]:
#Additional cleaning step: Changing inconsistent value of age and change datatype to int after doing that
student_df['age'] = student_df['age'].abs()
student_df['age'] = student_df['age'].astype("int64")
student_df['age']

0     17
1     16
4     19
5     20
6     17
7     21
8     17
10    16
11    17
13    19
14    20
15    18
16    16
17    17
18    18
19    19
20    20
21    21
Name: age, dtype: int64

# Part 3: Feature Engineering

## Create New Features

**1.1. Calculate total_score = math_score + reading_score**

In [11]:
student_df['total_score'] = student_df['math_score'] + student_df['reading_score']

#Check results
student_df[['math_score','reading_score','total_score']]

math_score  reading_score  total_score
0           85             78          163
1           92             88          180
4           78             81          159
5           88             90          178
6           65             70          135
7           95             92          187
8           72             68          140
10          82             85          167
11          89             76          165
13          93             87          180
14          77             73          150
15          70             90          160
16          68             82          150
17          84             79          163
18          91             88          179
19          74             75          149
20          99             95          194
21          60             85          145

**1.2. Convert exam_date to day_of_week (e.g., Monday)**

In [12]:
import datetime as dt

#Convert to dt datatype
student_df['exam_date'] = pd.to_datetime(student_df['exam_date'])

#Convert day to weekday using string formatted time
student_df['day_of_week'] = student_df['exam_date'].dt.strftime("%A")
student_df

student_id  age  gender  math_score  reading_score  attendance  \
0            1   17    Male          85             78      0.9500   
1            2   16  Female          92             88      0.8900   
4            5   19  Female          78             81      0.9100   
5            6   20   Other          88             90      0.9100   
6            7   17    Male          65             70      0.6800   
7            8   21  Female          95             92      0.8900   
8            9   17    Male          72             68      0.8355   
10          10   16  Female          82             85      0.9400   
11          11   17    Male          89             76      0.7500   
13          13   19  Female          93             87      0.7500   
14          14   20    Male          77             73      0.8900   
15          14   18   Other          70             90      0.7700   
16          15   16  Female          68             82      0.6800   
17          16   17    Male          84             79      0.9100   
18          17   18  Female          91             88      0.9400   
19          18   19    Male          74             75      0.8200   
20          19   20   Other          99             95      0.8900   
21          20   21  Female          60             85      0.7500   

   extracurricular  exam_date  total_score day_of_week  
0              Yes 2023-09-10          163      Sunday  
1              Yes 2023-09-11          180      Monday  
4               No 2023-09-10          159      Sunday  
5              Yes 2023-09-11          178      Monday  
6               No 2023-09-12          135     Tuesday  
7              Yes 2023-09-13          187   Wednesday  
8               No 2023-09-10          140      Sunday  
10             Yes 2023-09-11          167      Monday  
11              No 2023-09-12          165     Tuesday  
13              No 2023-09-10          180      Sunday  
14             Yes 2023-09-11          150      Monday  
15             Yes 2023-09-13          160   Wednesday  
16              No 2023-09-12          150     Tuesday  
17             Yes 2023-09-13          163   Wednesday  
18             Yes 2023-09-10          179      Sunday  
19              No 2023-09-11          149      Monday  
20             Yes 2023-09-12          194     Tuesday  
21              No 2023-09-13          145   Wednesday

## Binning

In [13]:
# Cut the Series into bins with data bin points and assign labels to the bins
student_df['age_group'] = pd.cut(x=student_df['age'], bins = [0,18,20, float('inf')], labels = ["<18", "18-20", ">20"])
student_df['age_group']

0       <18
1       <18
4     18-20
5     18-20
6       <18
7       >20
8       <18
10      <18
11      <18
13    18-20
14    18-20
15      <18
16      <18
17      <18
18      <18
19    18-20
20    18-20
21      >20
Name: age_group, dtype: category
Categories (3, object): ['<18' < '18-20' < '>20']

## Encode categorical Variables

**3.1. One-hot Coding on Gender**

In [14]:
#Initialize new columns with default values of 0
student_df['gender_Male'] = 0
student_df['gender_Female'] = 0
student_df['gender_Other'] = 0

#Update values to 1 if value in gender column equals to the corresponsing value of mask
student_df['gender_Male'] = student_df['gender_Male'].mask(student_df['gender'] == "Male", 1) 
student_df['gender_Female'] = student_df['gender_Female'].mask(student_df['gender'] == "Female", 1) 
student_df['gender_Other'] = student_df['gender_Other'].mask(student_df['gender'] == "Other", 1) 

#Check results
student_df[["gender", "gender_Male", "gender_Female", "gender_Other"]]

gender  gender_Male  gender_Female  gender_Other
0     Male            1              0             0
1   Female            0              1             0
4   Female            0              1             0
5    Other            0              0             1
6     Male            1              0             0
7   Female            0              1             0
8     Male            1              0             0
10  Female            0              1             0
11    Male            1              0             0
13  Female            0              1             0
14    Male            1              0             0
15   Other            0              0             1
16  Female            0              1             0
17    Male            1              0             0
18  Female            0              1             0
19    Male            1              0             0
20   Other            0              0             1
21  Female            0              1             0

**3.2. Label Coding on Extracurricular**

In [15]:
#Label-coding (Yes=1, No=0), filtering rows with extracurricular equals to the corresponsing value
student_df.loc[student_df['extracurricular'] == "Yes", "extracurricular"] = 1
student_df.loc[student_df['extracurricular'] == "No", "extracurricular"] = 0

#Check results
student_df['extracurricular']

0     1
1     1
4     0
5     1
6     0
7     1
8     0
10    1
11    0
13    0
14    1
15    1
16    0
17    1
18    1
19    0
20    1
21    0
Name: extracurricular, dtype: object

## Questions & Answers

4.1. Label encoding is a technique that directly modifies category values by assigning them numerical labels, whereas one-hot encoding separates each category into binary variables and assigns a binary value (0 or 1) based on its presence. Label encoding can assign values beyond 0 and 1 (0,1,2,3,4...), whereas one-hot encoding is strictly limited to binary values.

4.2. I will use these two techniques depending on the type of data. Label encoding may be more optimal as it saves memory, prevents an increase in DataFrame dimensions, and works well for ordinal data (e.g., mapping 1-5 for "Unsatisfied" to "Satisfied"). However, for nominal data, assigning integers in sequence can create a misleading sense of ranking. Therefore, in cases where the number of categories is low (such as gender or marital status), one-hot encoding remains a reasonable choice, as the additional columns are manageable while preserving data integrity.

# Part 4: Advanced Operations

## Broadcasting with Numpy

In [22]:
#Since the problem statement does not provide information, assume that the population size (N) is equal to the dataset size (N)
#Hence, we use NumPy to calculate the population standard deviation.
math_mean = np.mean(student_df['math_score'])
math_std = np.std(student_df['math_score'])

# z = (x - mean)/ std
student_df['math_zscore'] = (student_df['math_score'] - math_mean)/math_std
student_df['math_zscore']

0     0.344455
1     0.982710
4    -0.293800
5     0.617993
6    -1.479131
7     1.256248
8    -0.840876
10    0.070917
11    0.709172
13    1.073889
14   -0.384979
15   -1.023234
16   -1.205593
17    0.253276
18    0.891531
19   -0.658517
20    1.620965
21   -1.935027
Name: math_zscore, dtype: float64

## Time-series handling

In [17]:
#days between = targeted date - exam_date
current_date = pd.to_datetime("2025-01-31")
student_df['days_since_exam'] = current_date - student_df['exam_date']

#Check results
student_df[['exam_date','days_since_exam']]

exam_date days_since_exam
0  2023-09-10        509 days
1  2023-09-11        508 days
4  2023-09-10        509 days
5  2023-09-11        508 days
6  2023-09-12        507 days
7  2023-09-13        506 days
8  2023-09-10        509 days
10 2023-09-11        508 days
11 2023-09-12        507 days
13 2023-09-10        509 days
14 2023-09-11        508 days
15 2023-09-13        506 days
16 2023-09-12        507 days
17 2023-09-13        506 days
18 2023-09-10        509 days
19 2023-09-11        508 days
20 2023-09-12        507 days
21 2023-09-13        506 days

## Interaction Feature

In [18]:
#math_attendance = math_score * attendance
student_df['math_attendance'] = student_df['math_score'] * student_df['attendance']
student_df['math_attendance']

0     80.750
1     81.880
4     70.980
5     80.080
6     44.200
7     84.550
8     60.156
10    77.080
11    66.750
13    69.750
14    68.530
15    53.900
16    46.240
17    76.440
18    85.540
19    60.680
20    88.110
21    45.000
Name: math_attendance, dtype: float64

In [19]:
student_df

student_id  age  gender  math_score  reading_score  attendance  \
0            1   17    Male          85             78      0.9500   
1            2   16  Female          92             88      0.8900   
4            5   19  Female          78             81      0.9100   
5            6   20   Other          88             90      0.9100   
6            7   17    Male          65             70      0.6800   
7            8   21  Female          95             92      0.8900   
8            9   17    Male          72             68      0.8355   
10          10   16  Female          82             85      0.9400   
11          11   17    Male          89             76      0.7500   
13          13   19  Female          93             87      0.7500   
14          14   20    Male          77             73      0.8900   
15          14   18   Other          70             90      0.7700   
16          15   16  Female          68             82      0.6800   
17          16   17    Male          84             79      0.9100   
18          17   18  Female          91             88      0.9400   
19          18   19    Male          74             75      0.8200   
20          19   20   Other          99             95      0.8900   
21          20   21  Female          60             85      0.7500   

   extracurricular  exam_date  total_score day_of_week age_group  gender_Male  \
0                1 2023-09-10          163      Sunday       <18            1   
1                1 2023-09-11          180      Monday       <18            0   
4                0 2023-09-10          159      Sunday     18-20            0   
5                1 2023-09-11          178      Monday     18-20            0   
6                0 2023-09-12          135     Tuesday       <18            1   
7                1 2023-09-13          187   Wednesday       >20            0   
8                0 2023-09-10          140      Sunday       <18            1   
10               1 2023-09-11          167      Monday       <18            0   
11               0 2023-09-12          165     Tuesday       <18            1   
13               0 2023-09-10          180      Sunday     18-20            0   
14               1 2023-09-11          150      Monday     18-20            1   
15               1 2023-09-13          160   Wednesday       <18            0   
16               0 2023-09-12          150     Tuesday       <18            0   
17               1 2023-09-13          163   Wednesday       <18            1   
18               1 2023-09-10          179      Sunday       <18            0   
19               0 2023-09-11          149      Monday     18-20            1   
20               1 2023-09-12          194     Tuesday     18-20            0   
21               0 2023-09-13          145   Wednesday       >20            0   

    gender_Female  gender_Other  math_zscore days_since_exam  math_attendance  
0               0             0     0.344455        509 days           80.750  
1               1             0     0.982710        508 days           81.880  
4               1             0    -0.293800        509 days           70.980  
5               0             1     0.617993        508 days           80.080  
6               0             0    -1.479131        507 days           44.200  
7               1             0     1.256248        506 days           84.550  
8               0             0    -0.840876        509 days           60.156  
10              1             0     0.070917        508 days           77.080  
11              0             0     0.709172        507 days           66.750  
13              1             0     1.073889        509 days           69.750  
14              0             0    -0.384979        508 days           68.530  
15              0             1    -1.023234        506 days           53.900  
16              1             0    -1.205593        507 days           46.240  
17           

# Part 5: Short reflection

**5.1**

One challenge I encountered while working on this task was calculating the Z-score for math scores using NumPy and Pandas. Both Pandas and NumPy provide .mean() and .std() functions, but they handle standard deviation differently. To minimize risks, I calculated the Z-score using two approaches: one with parameters from NumPy and another with parameters from Pandas. The results showed a slight difference in the Z-scores (see below).

Through further research, I discovered that Pandas' .std() computes the sample standard deviation (dividing by N-1), while NumPy's .std() calculates the population standard deviation (dividing by N). Since the problem statement did not specify whether the dataset represents a sample or a population, my solution was to make an assumption about the population size. However, in real-world scenarios, this assumption should be explicitly validated to avoid introducing errors.

In [20]:
#Compute mean, standard deviation and zscore with numpy
mean_np = np.mean(student_df_copy['math_score'])
std_np = np.std(student_df_copy['math_score'])
student_df_copy['zscore_np'] = (student_df_copy['math_score'] - mean_np)/std_np

#Compute mean, standard deviation and zscore with pandas
mean_pd = student_df_copy['math_score'].mean()
std_pd = student_df_copy['math_score'].std()
student_df_copy['zscore_pd'] = (student_df_copy['math_score'] - mean_pd)/std_pd

#Compute differences between 2 zscore
student_df_copy['error'] = student_df_copy['zscore_np'] - student_df_copy['zscore_pd']

#Check results
student_df_copy[['zscore_np','zscore_pd', 'error']]

zscore_np  zscore_pd     error
0    0.344455   0.334750  0.009705
1    0.982710   0.955022  0.027688
4   -0.293800  -0.285522 -0.008278
5    0.617993   0.600581  0.017412
6   -1.479131  -1.437457 -0.041674
7    1.256248   1.220853  0.035394
8   -0.840876  -0.817184 -0.023691
10   0.070917   0.068919  0.001998
11   0.709172   0.689191  0.019981
13   1.073889   1.043633  0.030256
14  -0.384979  -0.374133 -0.010847
15  -1.023234  -0.994405 -0.028829
16  -1.205593  -1.171626 -0.033967
17   0.253276   0.246140  0.007136
18   0.891531   0.866412  0.025119
19  -0.658517  -0.639964 -0.018554
20   1.620965   1.575295  0.045670
21  -1.935027  -1.880508 -0.054519

**5.2.**

- Calculate total_score: To assess the overall performance of the students, examining whether there are consistently high-performing or low-performing students, or if the general performance is stable.
- Convert exam_date to days_of_the_week: To investigate whether the day of the week affects student performance and attendance.
- Binning into 3 age groups: To analyze the demographic details of the observations by classifying the dataset instead of treating age as a continuous variable.
- Encoding steps: Standardizing non-numeric variables into formats that machines can understand. This makes the dataset more flexible for aggregation calculations.